### PR TITLE
 Deduplicate memcpy, memmove, memset and memcmp functions

### PR DESCRIPTION
**Pull request by jaje**
Since `core::mem::size_of` is constant function there is no need for them to be duplicated.

### CODE REVIEW

**Comment by jackpot51** [comment]
Cool! Have you tested booting up with this change?

**Comment by jaje** [comment]
I tried 64bit build and it booted, I am going to try 32bit too.

**Comment by jackpot51** [comment]
Sweet! It looks good to me.

File: src/externs.rs [hunk 1]
```diff
@@ -57,84 +37,37 @@ pub unsafe extern fn memcpy(dest: *mut u8, src: *const u8,
...
+        let n_usize: usize = n/WORD_SIZE; // Number of word sized groups
         let mut i: usize = 0;
 
-        // Copy 4 bytes at a time
-        while i < n_32 {
-            *((dest as usize + i) as *mut u32) =
-                *((src as usize + i) as *const u32);
-            i += 4;
+        // Copy `WORD_SIZE` bytes at a time
+        while i < n_usize {
```
Comment (le-jzr) [inline src/externs.rs]: I think this (and similar sites in the rest of the file) should be `i <= n - WORD_SIZE`.
This way, you are copying only a fraction of the length here, and copy a byte at a time for almost all of it.

Comment (jackpot51) [reply src/externs.rs]: Good point. This would be a bug in the previous code as well.

Comment (le-jzr) [reply src/externs.rs]: Oh, it would also need an extra special case for `n < WORD_SIZE`, since they are unsigned. Unsigned ints are evil.

Comment (le-jzr) [reply src/externs.rs]: Edited my suggestion, the original one had potential overflow in it.

**Pull request merged**